### PR TITLE
Expected Consensus Update

### PR DESCRIPTION
**Pull request by sternhenri**
- Added a split out part about tipsets and how they function
- Added an actual bound for block delay
- Added lookback parameter construction update

### CODE REVIEW

File: expected-consensus.md [hunk 1]
```diff
@@ -1,104 +1,132 @@
...
 
-`ParentWeight` is the aggregate chain weight of a given block's parent set. It is calculated as
-the `ParentWeight` of any of its parent blocks (all blocks in a given `TipSet` should have 
-the same `ParentWeight` value) plus the delta weight of each parent. To make the 
-computation a bit easier, we store a block's parent weight in the block itself (otherwise 
-potentially long chain scans would be required to compute a given block's weight).
+## Secret Leader Election
 
-When selecting between tipsets of equal weight, choose the one with the smallest ticket (by bytewise comparison).
+Expected Consensus is a Nakamoto-style consensus protocol that works by electing a miner from a weighted set in proportion to their weight. In the case of Filecoin, participants and weights are drawn from the storage power table.
```
[the remark this answers is not in the document]
Comment (sternhenri) [reply expected-consensus.md]: mmh. insofar as it's incentivized probabilistic consensus I'd say it's nakamoto-style. That said, happy to remove it

File: expected-consensus.md [hunk 2]
```diff
@@ -1,104 +1,132 @@
...
+- The miner will use this ticket as input to a Verifiable Random Function, thereby ensuring secrecy: no other participant can generate this output without the miner's private key.
+- The miner will compare the output to their power from block N-L, with L the committee lookback parameter. If it is smaller, they have won and can mine a block inserting this winning ticket in the `ElectionProofs` array in the block. Else they wait to hear of another block generated in this round.
+
+We have at round N:
+
 ```
+T = Sig(H(sort(lookbackTickets)[0]))
 
-In English, that is the signature (using the miner's keypair) over the hash of the smallest ticket in the parent set. Note that this signature must be deterministic.
+Sig: Signature with the miner's keypair
```
[the remark this answers is not in the document]
Comment (sternhenri) [reply expected-consensus.md]: Apologies. No actually, in my construction there is no need to use a VDF here.

File: expected-consensus.md [hunk 3]
```diff
@@ -110,13 +138,57 @@ if winning(T) {
...
+
+Likewise, a miner should take their losing ticket from the original mining attempt (drawn from `K` blocks back), add it to the `ElectionProofs` array in the block and run a VRF on it once more, generating a new ticket to compare with their power in the table N-L blocks back. Each time it is discovered that nobody has won a given round, every miner should use their failed ticket to repeat the leader election process, appending said ticket to their would-be block's `ElectionProofs array. Once a miner finds a winning ticket, they can publish a block (see `Block Generation`).
 
-Thus, our ticket validation algorithm checks that the last ticket is a winning ticket and was adequately generated either from the parent set or from previous failed tickets.
+This new block (with multiple tickets in each array) will have a few key properties:
 
-## Slashing
+- All tickets in each array are signed by the same miner -- to avoid grinding through out-of-band collusion between miners exchanging tickets.
+- There is the same number of tickets in both arrays -- to ensure a miner has waited appropriately (in the `Ticket` array) before using their winning ticket (in the `ElectionProofs` array).
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply expected-consensus.md]: whats the VRF grinding they could do? the way i look at it, each additional attempt they make adds another null block, thus decreasing the chances that their block is accepted

[the remark this answers is not in the document]
Comment (sternhenri) [reply expected-consensus.md]: I believe we’re thinking of the same thing (repeated VRF attempts). It’s just that, to my understanding, additional null blocks don’t decrease chances a block is accepted. 

Currently this is solved by ensuring they actually take longer to be mined. There are alternatives (eg around weighting) but this seemed most in line with the design in normal case.

**Comment by whyrusleeping** [comment]
@sternhenri we've got a merge conflict, mind rebasing?

**Comment by sternhenri** [comment]
Should have caught that! Done and thank you!

File: expected-consensus.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply expected-consensus.md]: a chain with more null blocks will have a lower weight. so the more null blocks you have, the lower your chances of actually being chosen as *the* chain

[the remark this answers is not in the document]
Comment (sternhenri) [reply expected-consensus.md]: reporting back offline convo here: this is not true by necessity but will be most often true in that null blocks take time to mine, so likely other chains will have added true blocks in interim

File: expected-consensus.md [hunk 4]
```diff
@@ -1,121 +1,236 @@
...
+electionProof = Sig(H(sort(lookbackTickets)[0]))
+
+Sig: Signature with the miner's keypair, used as a VRF.
+H: Cryptographic compression function
+sort: bytewise sort
+lookbackTickets: Winning tickets from each of the `Tipset` blocks at round N-K
+```
 
-## Block Generation
+It is important to note that a miner generates two artifacts: one, a ticket from derived from last block's ticket to prove that they have waited the appropriate delay, and two, an election proof from K blocks back (where K could be 1) to run leader election.
```
Comment (whyrusleeping) [inline expected-consensus.md]:
```suggestion
It is important to note that a miner generates two artifacts: one, a ticket derived from last block's ticket to prove that they have waited the appropriate delay, and two, an election proof from K blocks back (where K could be 1) to run leader election.
```

**Pull request merged**